rllib/algorithms/dt/dt_torch_policy.py

```diff
@@ -0,0 +1,553 @@
+
+        act_shape = self.action_space.shape
+        act_dtype = self.action_space.dtype
+
+        # Here we will pad all the required inputs to its proper sequence length
+        # as their ViewRequirement.
+
+        observations = np.concatenate(
+            [
+                np.zeros((self.max_seq_len - 1, *obs_shape), dtype=obs_dtype),
+                observation[None],
+            ],
+            axis=0,
+        )
+
+        actions = np.zeros((self.max_seq_len - 1, *act_shape), dtype=act_dtype)
+
+        rtg = np.zeros(self.max_seq_len - 1, dtype=np.float32)
+
+        rewards = np.zeros((), dtype=np.float32)
+
+        # -1 for masking in action_distribution_fn
+        timesteps = np.full(self.max_seq_len - 1, fill_value=-1, dtype=np.int32)
+
+        input_dict = SampleBatch(
+            {
+                SampleBatch.OBS: observations,
+                SampleBatch.ACTIONS: actions,
+                SampleBatch.RETURNS_TO_GO: rtg,
+                SampleBatch.REWARDS: rewards,
+                SampleBatch.T: timesteps,
+            }
+        )
+        return input_dict
+
+    @PublicAPI
+    def get_next_input_dict(
+        self,
+        input_dict: SampleBatch,
+        action: TensorStructType,
+        reward: TensorStructType,
+        next_obs: TensorStructType,
+        extra: Dict[str, TensorType],
+    ) -> SampleBatch:
+        """Returns a new input_dict after stepping through the environment once.
+
+        Args:
+            input_dict: the input dict passed into compute_single_action.
+            action: the (unbatched) action taken this step.
+            reward: the (unbatched) reward from env.step
+            next_obs: the (unbatached) next observation from env.step
+            extra: the extra action out from compute_single_action.
+                In this case contains current returns to go *before* the current
+                reward is subtracted from target_return.
+
+        Returns:
+            A new input_dict to be passed into compute_single_action.
+            The input_dict for inference: {
+                OBS: [max_seq_len, obs_dim] array,
+                ACTIONS: [max_seq_len - 1, act_dim] array,
+                RETURNS_TO_GO: [max_seq_len - 1] array,
+                REWARDS: scalar,
+                TIMESTEPS: [max_seq_len - 1] array,
+            }
+            Note the sequence lengths are different, and is specified as per
+            view_requirements. Explanations in action_distribution_fn method.
+        """
+        # creates a copy of input_dict with only numpy arrays
+        input_dict = tree.map_structure(convert_to_numpy, input_dict)
+        # convert everything else to numpy as well
+        action, reward, next_obs, extra = convert_to_numpy(
+            (action, reward, next_obs, extra)
+        )
+
+        # Shift observations
+        input_dict[SampleBatch.OBS] = np.concatenate(
+            [
+                input_dict[SampleBatch.OBS][1:],
```

Comment:
should this be `[1:]` or `[-self.max_seq_len + 1:]`? Do you want to check the dimensions inside this method? e.g. input_dict['obs'] has the right number of timesteps?

Reply:
It should be [1:] because we're doing a sliding window, essentially removing the oldest and adding the newer one. It would be good to add a check for the size of the arrays though.

Reply:
added the check